tests/EdgeToEdge/APIRoutes/UpdateDonorControllerTest.php

```diff
@@ -0,0 +1,222 @@
+use WMDE\Fundraising\Frontend\Tests\EdgeToEdge\WebRouteTestCase;
+use WMDE\Fundraising\Frontend\Tests\Fixtures\FixedTokenGenerator;
+use WMDE\Fundraising\Frontend\Tests\Fixtures\StoredDonations;
+use WMDE\Fundraising\Frontend\Tests\RebuildDatabaseSchemaTrait;
+
+/**
+ * @covers \WMDE\Fundraising\Frontend\App\Controllers\API\Donation\UpdateDonorController
+ */
+class UpdateDonorControllerTest extends WebRouteTestCase {
+
+	use RebuildDatabaseSchemaTrait;
+	use GetApplicationVarsTrait;
+
+	private const PATH = 'api/v1/donation/update';
+	private const CORRECT_UPDATE_TOKEN = 'b5b249c8beefb986faf8d186a3f16e86ef509ab2';
+	private const INVALID_UPDATE_TOKEN = '2ba905fe68e61f3a681d8faf689bfeeb8c942b5b';
+
+	public function testWhenCorrectPrivatePersonDataIsPosted_addressIsChanged(): void {
+		$this->modifyConfiguration( [ 'skin' => 'laika' ] );
+		$client = $this->createClient();
+		$factory = $this->getFactory();
+		$donation = $this->newStoredDonation( $factory );
+
+		$this->performRequest(
+			$client,
+			$this->newPrivateDonorData(),
+			$donation->getId(),
+			self::CORRECT_UPDATE_TOKEN,
+			self::CORRECT_UPDATE_TOKEN
+		);
+		$response = $client->getResponse();
+
+		$expectedDonorData = $this->newPrivateDonorData();
+
+		$this->assertSame( Response::HTTP_OK, $response->getStatusCode() );
+		$this->assertJsonResponse( $expectedDonorData, $response );
+	}
+
+	public function testWhenCorrectCompanyDataIsPosted_addressIsChanged(): void {
+		$this->modifyConfiguration( [ 'skin' => 'laika' ] );
+		$client = $this->createClient();
+		$factory = $this->getFactory();
+		$donation = $this->newStoredDonation( $factory );
+
+		$this->performRequest(
+			$client,
+			$this->newCompanyDonorData(),
+			$donation->getId(),
+			self::CORRECT_UPDATE_TOKEN,
+			self::CORRECT_UPDATE_TOKEN
+		);
+
+		$response = $client->getResponse();
+
+		$expectedDonorData = $this->newCompanyDonorData();
+
+		$this->assertSame( Response::HTTP_OK, $response->getStatusCode() );
+		$this->assertJsonResponse( $expectedDonorData, $response );
+	}
+
+	public function testGivenRequestWithoutParameters_resultIsNotFound(): void {
+		$client = $this->createClient();
+
+		$client->jsonRequest(
+			Request::METHOD_PUT,
+			self::PATH,
+			[]
+		);
+
+		$response = $client->getResponse();
+
+		$this->assertSame( Response::HTTP_NOT_FOUND, $response->getStatusCode() );
+		$this->assertJsonResponse( [
+			'ERR' => 'No route found for "PUT http://localhost/api/v1/donation/update"'
+		], $response );
+	}
+
+	public function testWhenInvalidUpdateTokenIsSupplied_requestIsDenied(): void {
+		$client = $this->createClient();
+		$factory = $this->getFactory();
+		$donation = $this->newStoredDonation( $factory );
+
+		$this->performRequest(
+			$client,
+			$this->newPrivateDonorData(),
+			$donation->getId(),
+			self::CORRECT_UPDATE_TOKEN,
+			self::INVALID_UPDATE_TOKEN
+		);
+
+		$response = $client->getResponse();
+
+		$this->assertSame( Response::HTTP_BAD_REQUEST, $response->getStatusCode() );
```

Comment:
Technically, this should be a "403 - Access Denied". But we don't gain much value from checking this specific error result from the use case, so I'm inclined to approve this.

Reply:
Also, the front end probably won't ever care about the response other than if they're 200 or not. I think maybe 400 as a general code to cover all failure outcomes is acceptable?